enhancements/dns/forward-dns-over-tls.md

```diff
@@ -0,0 +1,334 @@
+DNS request forwarding improves the current security posture for DNS on 
+OpenShift.
+
+### Goals
+
+1. A cluster administrator can specify a CA certificate or CA bundle, TLS configuration, and an upstream TLS Server Name for the default zone in the DNS configuration.
+2. A cluster administrator can specify a CA certificate or CA bundle, TLS configuration, and an upstream TLS Server Name for a non-default zone.
+3. A cluster administrator can opt-in to DNS-over-TLS (using the configured CA/bundle) for each listed upstream in the default zone.
+4. A cluster administrator can opt-in to DNS-over-TLS (using the configured CA/bundle) for a set of upstreams on a non-default zone.
+
+### Non-Goals
+
+This enhancement applies only to the CoreDNS Operator shipped as part of OpenShift. Anything outside that scope would be not applicable to this enhancement including the CoreDNS static pods managed by the `machine-config-operator`.
+
+## Proposal
+
+This proposal covers changes to two APIs and the addition of the `DNSTransport` 
+type. 
+
+```go
+// DNSTransport is indicates what type of connection should be used
+// +kubebuilder:validation:Enum=tls;cleartext
+type DNSTransport string
+
+const (
+	// TLSTransport indicates that TLS should be used for the connection
+	TLSTransport DNSTransport = "tls"
+	
+	// CleartextTransport indicates that no encryption should be used for 
+	// the connection.
+	CleartextTransport DNSTransport = "cleartext"
+)
+```
+
+
+To enable a cluster administrator to configure DNS-over-TLS, the 
+`UpstreamResolvers` type will get three new optional members:
+
+`Transport` which can be either `TLSTransport` or `CleartextTransport`.
+
+`CABundle` references a ConfigMap that contains a single CA certificate 
+or a CA bundle.
+
+`ServerName` contains an FQDN to validate when establishing the connection. 
+This is relevant in particular with cases where a customer has multiple 
+upstreams behind one serving certificate (e.g. `1.1.1.1` and `2.2.2.2` 
+behind `dns.myorg.com`) as well as indicating what server we want to connect to.
+
+```go
+type UpstreamResolvers struct {
+  // ...<snip>
+  
+  // transport allows cluster administrators to opt-in to using a DNS-over-TLS 
+  // connection between cluster DNS and an upstream resolver(s). Configuring 
+  // TLS as the transport at this level without configuring a CABundle will 
+  // result in the system certificates being used to verify the serving 
+  // certificate of the upstream resolver(s).
+  //
+  // +optional
+  // +kubebuiler:default:="cleartext"
+  Transport DNSTransport `json:"transport,omitempty"`
+  
+  // caBundle references a ConfigMap that must contain either a single
+  // CA Certificate or a CA Bundle (in the case of multiple upstreams signed
+  // by different CAs). This allows cluster administrators to provide their 
+  // own CA or CA bundle for validating the certificate of upstream resolvers.
```

Comment:
describe what happens when unset and transport is TLS.  can be done in the API PR instead of here.

Reply:
ack, will do!